### PR TITLE
WIP: a big 'ol refactor and new features

### PR DESCRIPTION
**Pull request by kodonnell**
This was largely based off discussions in #4 and #3 (including you `issues/3` as a starter), except I went the whole hog (as below). This is nearly complete feature-wise, but needs a bunch of tidying. Key things to note:

- Generally more detailed field definitions/values. For example, enumerated fields are `Enum`s, and bitfields are `IntFlag`s, and each field type has it's own class, with it's own validations (e.g. exceeding max specified in SunSpec) etc. 
- This is now a thing as per #3 :
  ```python
  # get absorb volts for charge controller on port 3
  volts = client.devices.charge_controllers[3].config.absorb_volts
  # or, if there's only one
  volts = client.devices.charge_controller.config.absorb_volts
  ```
- But you get more info about fields ...
  ```python
  # when was it last read?
  volts.last_read
  # is it dirty (i.e. has it been assigned?)
  volts.dirty
  # is it implemented?
  volts.implemented
  # what's the value (taking into account scale factors)
  volts.value
  ```
- Reading individual fields is still a thing. It's arguably nicer in that the user never has to worry about the sunspec definition vs values:
  ```python
  # only read the CC's absorb volts (and automatically any scale_factors just in case)
  client.read(only=[client.devices.charge_controller.config.absorb_volts])
  ```
- Writing (still untested) is now nicer:
  ```python
  client.devices.charge_controller.config.absorb_volts = 57.6 # NB: we don't have to worry about scale factors
  client.write()
  ```
- Used the new XLSX, and changed the generator code to use that. Added in bitfield support from those tables. There are a few gross hacks in the code generation (including fixing seeming Outback errors e.g. not labeling some enumerated fields as enumerated), but since they're highly specific to the exact file format anyway, it's OK. We could let Outback know I guess.
- Oh, a caching modbus connection - you'll figure it out, but you run it once with `cache_only=False` and that caches the modbus lookups, and you can then run again with `cache_only=True` and it won't hit the mate ... which was nice for not bricking it as I debugged my code! I might tidy this up for re-use in testing etc.
- I'm still not 100% on my SunSpec terminology e.g. models vs blocks etc.
- Oh, in `FXInverterConfigurationModel`, the values I get for `input_support` (54 - though there are some comments about validity in the desc, but they don't seem to apply here) and `grid_tie_enable` (255) are invalid for the enums specified. Not sure what to do there.
- The key files are:
  - `mate3/sunspec/fields.py` - generic definitions about sunspec'y fields.
  - `mate3/sunspec/scripts/code_generator.py` - same as before, but on the XLSX
  - `mate3/sunspec/models.py` - the generated sunspec model definitions parsed from the XLSX.
  - `mate3/sunspec/values.py` - same as the old 'structures.py' - this is just a definition of a container per model for values that are read. This is actually redundant, and was being done dynamically off the model definition, but then you don't get nice code completion (I don't think?) so I generated it.
  - `mate3/api.py` - standard stuff for reading etc., but also the dynamic reflection of a model definition as an object to store data in.
  - `mate3/devices.py` - this maps e.g. both the charge controller models into the `client.devices.charge_controllers` attribute.

It's not complete (including the mate3 CLI), but `api.py` works and prints stuff on my mate3 (not mate3s) with latest firmware. Keen for some initial thoughts @adamcharnock if you have time? ~~Please note that a lot of the 'minor' details I haven't worried about e.g. the exact names of things, whether fields should be accessed by an enum instance or the `enum.name`, exactly how to serve up the field values to the user (at the moment it's just a dict for debugging) etc. I also haven't implemented some things e.g. the writing of values, or reading only particular values etc. - but they should be 'straightforward' and not involve significant refactors.~~

On this PR - apologies that it's one massive commit, but committing all of the rabbit holes I went down while refactoring wouldn't have been fun. It's totally OK to say "woah there, I'm not going to merge something this massive" - I've sunk enough hours into this already (part of it learning some cool things from your code - thanks!), and as the original library would've worked fine for my (basic) needs of monitoring only a few values, I think I'm happy to leave this here. (I could also potentially split this into multiple PRs if you wish.) Also, if you could ignore any opinion stuff for now (e.g. my TODO on ignoring data types) that'd be good - I'm not going to die on any of those hills, and we can get to them if you're happy with the general overview. Likewise for obviously bad/temporary code - I'll tidy all that up later.

Other random thoughts:
- PySunSpec seems to do a lot of this, but doesn't seem particularly well maintained, roll their own modbus, and doesn't have the mate3 specific definitions/apis etc.
- Haven't tested the FNDC issue you saw with currents yet. Haven't compared most fields in detail, except a quick eyeball of the obvious ones like voltages.

### CODE REVIEW

**Comment by kodonnell** [comment]
@adamcharnock I've updated this a bit and it's getting closer to feature complete aside from tidyups and finishing `client.write`. Oh, and reading only some fields. Do you have any initial thoughts? I don't need a full code review, just whether e.g. you like the general API changes and you think it's worth polishing up.

**Comment by kodonnell** [comment]
@adamcharnock just wondering if you'd had time to review? I've also been swamped, but wouldn't mind getting some direction on whether I should continue with this ...

**Comment by adamcharnock** [comment]
Hi @kodonnell, this looks absolutely fantastic, thank you so much for submitting this! 😄 👍 

I'm very happy for you to proceed in this way. I'd also be happy to make you a maintainer of this repo too if that's ok with you. I think this PR qualifies you for it, and I don't want you to have to keep waiting around for a month for me to reply!

**Comment by kodonnell** [comment]
Awesome, thanks for the feedback @adamcharnock - I'll plug away tidying up all the loose ends, and throw some tests around etc. And yes, maintainer would be great. Some more questions (while I've got you!):

- Do you want to review the final PR or just have me merge it?
- Do you want me to update PyPi? I've never actually put anything on PyPi before, but if you can somehow make me maintainer too, I can do that for you. Otherwise I'll leave it to you.

**Comment by adamcharnock** [comment]
> Do you want to review the final PR or just have me merge it?

Give me a few days to review it, and if you don't hear from me just go ahead and merge.

> Do you want me to update PyPi?

Sure, send me your PyPi username. The release process is in the README.

**Comment by kodonnell** [comment]
Sounds good - when it's ready for review I'll let you know. PyPi username is `kodonnell`.

**Comment by kodonnell** [comment]
Getting closer! Mostly just code review / tidying now (inc. tests? though might leave them to another PR), and documentation/deployment stuff, and updating the CLI.

**Comment by adamcharnock** [comment]
Amazing!

I'm happy for tests to go in another PR, given that there are zero tests right now anyway.

**Comment by chinezbrun** [comment]
great job here guys! amazing!  when will have the PyPi updated?

**Comment by kodonnell** [comment]
Glad to know you're keen @chinezbrun! I've made a bit more progress (e.g. the mate3 command now works) but would prefer to finish code tidyups (about half done) and documentation before pushing to PyPi. In the interim, can you try this?

```
pip install git+https://github.com/kodonnell/mate3
```

If you run the mate3 command, I'd be interested to see if it works on your hardware (as I don't have radians) and fixes your bug from the other issue - and any other comments. NB: I haven't tested writing yet, but I think it should work - you're welcome to try and let me know = ) There's an example usage of the API at the bottom of `api.py` - though if you're setting bitfields and stuff you might need to do it with intflags and stuff (which I maybe wanna make more transparent, but will do for now).

**Comment by chinezbrun** [comment]
[mate3_Modbus_all_blocks - Copy.txt](https://github.com/adamcharnock/mate3/files/4941455/mate3_Modbus_all_blocks.-.Copy.txt)

I'm using attached code to extract in json format all data from Mate3S. Anything to be changed?

**Comment by kodonnell** [comment]
@adamcharnock I think this is ready for review - at least all the major stuff. Key work remaining:

 - Decide what to do about `mate3_pg.py` given it's more of an example use as opposed to core to the code. Maybe we move it into ./examples ?
- Test that write actually works on my hardware! I'll just change the name etc. I think. 
- Release etc.

But I think you can have a review of the code as-is now - it's probably clean enough, but not polished. I'm probably not going to be able to get to this until next weekend until the earliest, so I'll wait until then to hear from you. Feel free to throw opinions round (e.g. "I don't like loguru") - I won't be offended. Also - it's been cool refactoring this, as I learned a bunch of python tricks from your code.

Finally, here's a list of future work that I might turn into issues (though I have no immediate attention of working on them):

- Allow reading only a single field *from a single device* (i.e. a device on a specific port if there are multiple). We'd still need to read all devices of this type to get the port, but we could avoid further reads on devices we don't care about. Probably a micro-optimization.
- More logging etc.
- For write in CLI:
  - `eval`, yuck! Will need a `from_string` method for all field types to handle this.
  - It's pretty yuck for enums/bitfields e.g. `--set="charge_controller.config.grid_tie_mode=client.devices.charge_controller.config.grid_tie_mode.field.options['Grid Tie Mode disabled']"`. Probably not too much we can do about it.
- Optimise reading `only` fields so that once we've got what we need, don't continue reading all models (which we current do for address/length). Note that we'd need to change the device `update` method, as it currently assumes we're getting something from all devices (as it'll e.g. drop devices which it doesn't receive data from).
- Tests!
  - Integration - get a few modbus caches from people.
  - Test all fields / field values - conversion to/from modbus etc., scale factors, etc.
  - Test the state updates over time with various reads/writes/etc.
  - ...
- Remove passwords from being cached in the debug?
- Figure these out:
> 2020-07-19 21:27:57.782 | WARNING  | mate3.api:_read_model:222 - Error reading field input_support - so setting as not implemented. Message: 54 is not a valid input_support
> 2020-07-19 21:27:57.788 | WARNING  | mate3.api:_read_model:222 - Error reading field grid_tie_enable - so setting as not implemented. Message: 255 is not a valid grid_tie_enable
- Better naming ...
  - Fields ... not too bad, but it conflicts with dataclass fields, and is maybe too general. Data point is the modbus thing, IIRC.
  - FieldValues/values - these are like the previous 'structures' - just containers for data corresponding to a given field. Maybe 'data' is more appropriate? I'm just generally wary that 'values' is too generic which can make code conflicts a thing.
  - Device/Model/ModelValues - not sure if I'm using the SunSpec terms correctly, but yeh, check this.
- A bit more thinking on how to handle situations like switching the ports on two inverters while the same python process is running. E.g. if port 1 had a value to write of X (but we hadn't called client.write), then we rewire, will the write now happen on the device at port 1 or port 2? (And how do we make it do what we want?)

**Comment by kodonnell** [comment]
@chinezbrun That's the old API. If you want python, you can check out the `examples/getting_started.py` example in this PR. However, the CLI makes sense here. If you use bash:

```bash
while true; do
    mate3 read --host=192.168.0.150 --format=json > "mate3.json"
    sleep(3)
done
```

I haven't tested, but that should roughly be it. Let me know if it fixes #9 !

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Outback Mate 3s Python library & command line interface
+# Outback Mate 3 & 3s Python library & command line interface
```
Comment (adamcharnock) [inline README.md:1]: I'm not sure this actually works on the Mate3. I think I just forgot the `s` when I wrote the original readme, so this has only ever been run against the Mate3s. Just mentioning it so you are aware, not sure any change is required. Maybe it *does* work with the Mate3 🤷‍♂️

File: README.md [hunk 2]
```diff
@@ -232,6 +205,5 @@ git push --tags
 
 ## Credits
 
-This is a heavily refactored version of 
-[basrijn's Outback_Mate3 library](https://github.com/basrijn/Outback_Mate3).
-Thank you basrijn!
+This was originally a heavily refactored version of
+[basrijn's Outback_Mate3 library](https://github.com/basrijn/Outback_Mate3), though has largely been completely rewritten since. Thanks anyway basrijn!
```
Comment (adamcharnock) [inline README.md:209]: Just FYI – I'll have a read through this once-merged (and therefore easier to read) and may make some tweaks

File: mate3/api.py [hunk 3]
```diff
@@ -164,54 +303,56 @@ def set_value(self, field: Field, value: int, port: int = None):
...
+                    address = device.address + field_value.field.start - 1  # -1 since start is 1-indexed
+                    # TODO: should we check that the device at the given address still has the right port?
+                    registers = field_value.field.to_registers(field_value._value_to_write)
+                    logger.debug(f"Setting register {address} to value {registers}")
+
+                    # # Do the write
+                    self.client.write_registers(address, registers)
+
+
+if __name__ == "__main__":
```
Comment (adamcharnock) [inline mate3/api.py:324]: This looks like debug code, perhaps this needs to be removed or placed elsewhere such as a script entrypoint?

File: mate3/api.py [hunk 4]
```diff
@@ -164,54 +303,56 @@ def set_value(self, field: Field, value: int, port: int = None):
...
+        print("# Cool. Can we set a new value? Note that we don't need to worry about scaling etc.")
+        volts = client.devices.charge_controller.config.absorb_volts
+        print("Before:", volts)
+        client.devices.charge_controller.config.absorb_volts.value = volts.value + 0.1
+        print("After: ", volts)
+        print("# OK, but what about fun fields like Enums?")
+        new_value = client.devices.charge_controller.config.grid_tie_mode.field.options["Grid Tie Mode disabled"]
+        client.devices.charge_controller.config.grid_tie_mode.value = new_value
+        print("# Finally, write any values that have changed to the device itself - BE CAREFUL!")
+        client.write()
```
Comment (adamcharnock) [inline mate3/api.py:358]: There is a lot going on here, so I haven't read through it in detail. Thanks for the comments and docstrings through

File: mate3/code_generator.py [hunk 5]
```diff
@@ -1,296 +0,0 @@
...
-    format_python_file(PARSERS_MODULE)
-
-    with open(str(STRUCTURES_MODULE), 'w') as f:
-        f.write(structure_code)
-
-    format_python_file(STRUCTURES_MODULE)
-
-
-if __name__ == '__main__':
-    main()
```
Comment (adamcharnock) [inline mate3/code_generator.py:296]: Is the code generation not needed any more?

File: mate3/csv_generator.py [hunk 6]
```diff
@@ -1,25 +0,0 @@
...
-        # Sometimes camelot will merge two cells together because it is
-        # trying to be smart. Don't allow it because it doesn't happen
-        # anywhere, and because it can happen in cells with close-together text
-        split_text=True,
-    )
-
-    for old_file in CSV_DESTINATION.glob('*.csv'):
-        old_file.unlink()
-
-    tables.export(CSV_DESTINATION / 'registry-data.csv', f='csv', compress=False)
```
Comment (adamcharnock) [inline mate3/csv_generator.py:25]: Fair enough that this is gone as we don't need to read PDFs any more

File: mate3/examples/getting_started.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        print("# Cool. Can we set a new value? Note that we don't need to worry about scaling etc.")
+        volts = client.devices.charge_controller.config.absorb_volts
+        print("Before:", volts)
+        client.devices.charge_controller.config.absorb_volts.value = volts.value + 0.1
+        print("After: ", volts)
+        print("# OK, but what about fun fields like Enums?")
+        new_value = client.devices.charge_controller.config.grid_tie_mode.field.options["Grid Tie Mode disabled"]
+        client.devices.charge_controller.config.grid_tie_mode.value = new_value
+        print("# Finally, write any values that have changed to the device itself - BE CAREFUL!")
+        client.write()
```
Comment (adamcharnock) [inline mate3/examples/getting_started.py:37]: Good idea on having an examples directory!

File: mate3/field_values.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            self._value_to_write = int(round(value, 0))
+        else:
+            self._value_to_write = value
+        self._dirty = True
+
+    def _update_on_read(self, value, implemented, read_time, scale_factor_read=None):
+        """
+        Assumption is that if there's a scale factor, it's read at the same time as value, so they should be in sync.
+        Not really a major with Outback, as the scale factors seem to be constant anyway ... but no harm in being safe.
+        """
```
Comment (adamcharnock) [inline mate3/field_values.py:122]: Minor comment – Can this docstring start by saying what the method does?

**Comment by adamcharnock** [comment]
This is great @kodonnell, and what a lot of work you've done! I'm working though it now.

Adding a comment here which I cannot add in the PR:

* Can you also/instead include a CSV version of `OutBack.Power.SunSpec.Map.xlsx`? Just in the name of open formats and all that :-)

File: mate3/code_generator.py (hunk 5 above)
Comment (adamcharnock) [reply mate3/code_generator.py:296]: Oh I see, it is moved elsewhere, never mind :-)

**Comment by adamcharnock** [comment]
Done!

> Decide what to do about mate3_pg.py given it's more of an example use as opposed to core to the code. Maybe we move it into ./examples ?

`mate3_pg` is my main use of this library. I use it to write data to postgres (timescaledb) every 3 seconds, which then provides my monitoring through grafana. I have a fairly string preference for it staying. How do you feel about this?

File: README.md (hunk 1 above)
Comment (kodonnell) [reply README.md:1]: Mine's a Mate3 so we've got both covered!

File: mate3/api.py (hunk 3 above)
Comment (kodonnell) [reply mate3/api.py:324]: Good point - I actually moved it out to `examples/getting_started.md` so I'll delete it from `api.py`.

File: mate3/api.py (hunk 4 above)
Comment (kodonnell) [reply mate3/api.py]: No problems. I tried to convey the various useful things, but it's really wanting a nice bit of documentation / FAQ.

File: mate3/field_values.py (hunk 8 above)
Comment (kodonnell) [reply mate3/field_values.py]: Done!

**Comment by kodonnell** [comment]
> Done!

Awesome, thanks.

> mate3_pg is my main use of this library. I use it to write data to postgres (timescaledb) every 3 seconds, which then provides my monitoring through grafana. I have a fairly string preference for it staying. How do you feel about this?

Fine with me - happy for me to leave it as a separate script? My general preference for when this becomes really popular (!) is that w'de keep this library focused on just the Mate3 API and put applications of it into new repos, in the same way you don't have grafana in here. But since we aren't quite that popular yet, I think we can have some flexibility = )

> I use it to write data to postgres (timescaledb) every 3 seconds, which then provides my monitoring through grafana

Interesting, and would love to see! This all started because I want to be able to monitor my system more effectively for two main cases: 1) general monitoring/alerting (hot inverters, decrease in turbine production == bearing failure / air locks, etc.) and ability to do basic things away-from-home when things go south (e.g. shut the inverters down), and 2) do some smart stuff with my excess power instead of burning it all off (which will require some local devices to control some SSRs on HWCs etc.). The former I'm planning on doing with something like Azure Iot Central or similar (partly as a learning exercise), and the latter with some local compute. Anyway - if you're interested to collaborate, let me know and I'll flick you a message.

**Comment by chinezbrun** [comment]
> If you run the mate3 command, I'd be interested to see if it works on your hardware (as I don't have radians) and fixes your bug from the other issue - and any other comments.

Mate3 command is running on Linux, not running on Windows (i was not able to install it :(
Tested with Radian. Mate3s, API is working well, see the json file. Unfortunately issue #9 is not fixed.
In attached file you can see that sched_2_ac_mode = 0 and sched_3_ac_mode = 0 in reality sched_2_ac_mode = Generator and sched_3_ac_mode = Disabled

**Comment by kodonnell** [comment]
Sorry about the delay! @chinezbrun I think that's the old mate3 API. If you run something like

```
mate3 read --host=192.168.0.150 -f json
```

In the new API you should see things like:

image

Whereas what you sent looks like:

image

If you want, you could also `mate3 debug` to send through your raw modbus values, and then I can use them for testing on my end. Just note that this includes everything - including passwords etc. if you have them. Though you've probably already sent that above.

**Comment by kodonnell** [comment]
@adamcharnock  Sorry, missed this one:

> Can you also/instead include a CSV version of OutBack.Power.SunSpec.Map.xlsx? Just in the name of open formats and all that :-)

I'm always wary of having two sources of truth, so I'd prefer to leave just the XLSX - I was actually a bit confused when I first saw this repo why there was a PDF and some CSVs. Also, it'd be some amount of work to properly convert XLSX -> CSV (since there are many tables etc.).

> [Re postgres, I said ... ] Fine with me - happy for me to leave it as a separate script? 

I didn't hear back from you on this, so I just did it! Are you OK with this?

Finally got back to this, so updates:

- Writing to the mate3 works! Well, I only tested `mate3.system_name`, but that'll do for now. (Oddly enough, `system_name` doesn't seem to change the name on the display, even though modbus values are changed when I read them ...)
- I've got it ready (?) for release. I haven't done this before (releasing packages to PyPI, let alone using poetry et. al.). @adamcharnock if you could have a quick look, that'd be great. (For reference, I stopped before `poetry publish --build` in the readme.) Also, I'm not well-versed in semver ... this feels like a major since it's breaking, but also, since we're still in 0.*, I figured a minor to 0.5.0 would be OK? If I haven't heard from you in the next few days (or when I get back to this!), I'll just go with what I've got.

**Comment by adamcharnock** [comment]
Hey @kodonnell – sorry for the radio silence on this. I've caught up now I think you have made good choices all around, so nice work. I haven't checked the code out myself yet, but hopefully I'll get a chance to update my systems to it soon.

**Comment by kodonnell** [comment]
> sorry for the radio silence on this

Hah - I also went dark for months! I figured it was best to just merge/publish this stuff, and get it out of the way, and any remaining changes/additions/fixes can come later.

>  I haven't checked the code out myself yet, but hopefully I'll get a chance to update my systems to it soon.

FWIW I'm planning on getting something sorted over the next month, as I've got a nice long holiday. This means I'll probably find and fix some bugs (that are no doubt there) once I start actually controlling my system with some writes etc.

**Pull request merged**